src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp

```diff
@@ -302,21 +307,65 @@ struct RemoveRecursiveObjectStorageOperation final : public IDiskObjectStorageOp
         IMetadataStorage & metadata_storage_,
         const std::string & path_,
         bool keep_all_batch_data_,
-        const NameSet & file_names_remove_metadata_only_)
+        const NameSet & file_names_remove_metadata_only_,
+        UInt64 limit_)
         : IDiskObjectStorageOperation(object_storage_, metadata_storage_)
         , path(path_)
         , keep_all_batch_data(keep_all_batch_data_)
         , file_names_remove_metadata_only(file_names_remove_metadata_only_)
+        , limit(limit_)
     {}
 
     std::string getInfoForLog() const override
     {
         return fmt::format("RemoveRecursiveObjectStorageOperation (path: {})", path);
     }
 
+    bool checkLimitReached() const
+    {
+        return limit > 0 && objects_to_remove_by_path.size() == limit;
+    }
+
+    void removeObjects()
+    {
+        if (!keep_all_batch_data)
```

Comment:
If `keep_all_batch_data` is true `objects_to_remove_by_path` will not be cleared